File: server/routers/playlists.py
"""Плейлисты и их содержимое."""
from fastapi import APIRouter, Body, Depends, HTTPException, Query
from sqlalchemy import text
from sqlalchemy.orm import Session

from deps import get_db, get_current_admin, require_write

router = APIRouter()


@router.post("/playlists")
def create_playlist(name: str = Query(...), admin: dict = Depends(require_write),
                    db: Session = Depends(get_db)):
    result = db.execute(text("INSERT INTO playlists (name) VALUES (:n) RETURNING id"), {"n": name})
    pid = result.fetchone()[0]
    db.commit()
    return {"id": pid, "name": name}


@router.get("/playlists")
def list_playlists(admin: dict = Depends(get_current_admin), db: Session = Depends(get_db)):
    rows = db.execute(text("SELECT * FROM playlists ORDER BY name")).fetchall()
    return [dict(r._mapping) for r in rows]


@router.post("/playlists/{playlist_id}/items")
def add_to_playlist(
    playlist_id: int,
    media_id: int = Query(...),
    repeat_count: int = Query(1),
    admin: dict = Depends(require_write),
    db: Session = Depends(get_db)
):
    # 38-ФЗ: неодобренный или отклонённый контент не попадает в эфирные плейлисты
    m = db.execute(text("SELECT title, review_status, status FROM media WHERE id = :id"),
                   {"id": media_id}).fetchone()
    if not m:
        raise HTTPException(404, "Ролик не найден")
    # Файлы из папки «Документы» (сканы договоров и деклараций) хранятся со
    # status='document' и модерацию не проходят — в эфирный плейлист их нельзя.
    if m.status == "document":
        raise HTTPException(409, f"«{m.title}» — документ, а не ролик: в плейлист добавить нельзя")
    if m.review_status != "approved":
        status_ru = {"pending": "ещё на модерации", "rejected": "отклонён модератором"}.get(
            m.review_status, m.review_status)
        raise HTTPException(409, f"Ролик «{m.title}» {status_ru} — в плейлист можно добавлять только одобренные ролики")

    # #1: один и тот же ролик нельзя добавить в плейлист дважды
    dup = db.execute(text(
        "SELECT 1 FROM playlist_items WHERE playlist_id = :pid AND media_id = :mid"
    ), {"pid": playlist_id, "mid": media_id}).fetchone()
    if dup:
        raise HTTPException(409, f"Ролик «{m.title}» уже есть в этом плейлисте")

    pos = db.execute(text(
        "SELECT COALESCE(MAX(position), 0) + 1 FROM playlist_items WHERE playlist_id = :pid"
    ), {"pid": playlist_id}).scalar()
    result = db.execute(text(
        "INSERT INTO playlist_items (playlist_id, media_id, position, repeat_count) "
        "VALUES (:pid, :mid, :pos, :rc) RETURNING id"
    ), {"pid": playlist_id, "mid": media_id, "pos": pos, "rc": repeat_count})
    db.commit()

    # #2: предупреждение, если ролик уже используется в других плейлистах
    others = db.execute(text(
        "SELECT p.name FROM playlist_items pi JOIN playlists p ON p.id = pi.playlist_id "
        "WHERE pi.media_id = :mid AND pi.playlist_id <> :pid ORDER BY p.name"
    ), {"mid": media_id, "pid": playlist_id}).fetchall()
    warning = None
    if others:
        names = ", ".join(f"«{o.name}»" for o in others)
        warning = f"Этот ролик уже используется в плейлисте(ах): {names}"

    return {"id": result.fetchone()[0], "playlist_id": playlist_id,
            "media_id": media_id, "position": pos, "warning": warning}


@router.get("/playlists/{playlist_id}/items")
def get_playlist_items(playlist_id: int, admin: dict = Depends(get_current_admin),
                       db: Session = Depends(get_db)):
    rows = db.execute(text("""
        SELECT pi.id, pi.position, pi.repeat_count,
               m.id as media_id, m.title, m.filename, m.filesize, m.duration_seconds
        FROM playlist_items pi JOIN media m ON m.id = pi.media_id
        WHERE pi.playlist_id = :pid ORDER BY pi.position
    """), {"pid": playlist_id}).fetchall()
    return [dict(r._mapping) for r in rows]


@router.patch("/playlists/{playlist_id}/fill")
def set_playlist_fill(playlist_id: int, body: dict = Body(...),
                      admin: dict = Depends(require_write), db: Session = Depends(get_db)):
    """Режим длительности плейлиста: fill_to_hour=true — эфирный блок
    фиксированной длины (target_minutes, по умолчанию 60), добивается
    заглушками; false — произвольный плейлист, играет как есть."""
    if not db.execute(text("SELECT 1 FROM playlists WHERE id=:id"), {"id": playlist_id}).fetchone():
        raise HTTPException(404, "Плейлист не найден")
    fill = bool(body.get("fill_to_hour", True))
    minutes = body.get("target_minutes", 60)
    try:
        minutes = int(minutes)
    except (TypeError, ValueError):
        raise HTTPException(400, "target_minutes: целое число минут")
    if not (1 <= minutes <= 24 * 60):
        raise HTTPException(400, "target_minutes: от 1 до 1440 минут")
    db.execute(text(
        "UPDATE playlists SET fill_to_hour=:f, target_seconds=:t WHERE id=:id"
    ), {"f": fill, "t": minutes * 60, "id": playlist_id})
    db.commit()
    return {"status": "ok", "fill_to_hour": fill, "target_minutes": minutes}


@router.get("/playlists/{playlist_id}/fill_info")
def playlist_fill_info(playlist_id: int, admin: dict = Depends(get_current_admin),
                       db: Session = Depends(get_db)):
    """Расчёт заполнения эфирного блока: сколько идут основные ролики, сколько
    заглушек добавится, и предупреждения (нет заглушек / не хватает / перебор)."""
    pl = db.execute(text(
        "SELECT COALESCE(fill_to_hour, TRUE) AS fill_to_hour, "
        "COALESCE(target_seconds, 3600) AS target_seconds "
        "FROM playlists WHERE id=:id"), {"id": playlist_id}).fetchone()
    if not pl:
        raise HTTPException(404, "Плейлист не найден")

    from admin_panel import build_hourly_detail
    d = build_hourly_detail(db, playlist_id, int(pl.target_seconds))

    warnings = []
    # Для пустого плейлиста агент не играет ничего (в т.ч. заглушки) —
    # предупреждать о нехватке заглушек бессмысленно
    if pl.fill_to_hour and d["main_plays"] > 0:
        target = float(pl.target_seconds)
        remaining = target - d["main_seconds"]
        if d["main_seconds"] > target:
            warnings.append(
                f"Основные ролики идут {d['main_seconds']/60:.1f} мин — это БОЛЬШЕ целевых "
                f"{target/60:.0f} мин. Блок будет длиннее цели, заглушки не добавляются. "
                "Уберите ролики или увеличьте целевую длительность.")
        elif remaining > 0 and d["fillers_available"] == 0:
            warnings.append(
                f"До {target/60:.0f} мин не хватает {remaining/60:.1f} мин, а заглушек в "
                "медиатеке нет — эфирный блок будет короче цели. Загрузите ролики в "
                "Медиатека → Заглушки.")
        elif remaining > 0 and d["fillers_available_seconds"] < remaining:
            warnings.append(
                f"Свободные {remaining/60:.1f} мин заполняются заглушками, но их суммарная "
                f"длительность всего {d['fillers_available_seconds']/60:.1f} мин — заглушки "
                "будут повторяться по кругу. Загрузите больше заглушек для разнообразия.")

    return {
        "fill_to_hour": bool(pl.fill_to_hour),
        "target_seconds": int(pl.target_seconds),
        "main_seconds": round(d["main_seconds"], 1),
        "main_plays": d["main_plays"],
        "filler_seconds": round(d["filler_seconds"], 1),
        "filler_plays": d["filler_plays"],
        "total_seconds": round(d["main_seconds"] + d["filler_seconds"], 1),
        "fillers_available": d["fillers_available"],
        "fillers_available_seconds": round(d["fillers_available_seconds"], 1),
        "warnings": warnings,
    }


@router.delete("/playlists/{playlist_id}")
def delete_playlist(playlist_id: int, admin: dict = Depends(require_write),
                    db: Session = Depends(get_db)):
    db.execute(text("DELETE FROM playlists WHERE id=:id"), {"id": playlist_id})
    db.commit()
    return {"status": "deleted"}


@router.delete("/playlists/{playlist_id}/items/{item_id}")
def delete_playlist_item(playlist_id: int, item_id: int,
                         admin: dict = Depends(require_write), db: Session = Depends(get_db)):
    db.execute(text("DELETE FROM playlist_items WHERE id=:id AND playlist_id=:pid"),
               {"id": item_id, "pid": playlist_id})
    db.commit()
    return {"status": "deleted"}

File: server/static/js/views/playlists.js
//=============================================================================
// ПЛЕЙЛИСТЫ
//=============================================================================

let CUR_PL = null;

function initPlaylistsViewActions(){
  if(window.__playlistsViewActionsInitialized) return;
  window.__playlistsViewActionsInitialized = true;

  document.addEventListener('click', e => {
    const el = e.target.closest('[data-action]');
    if(!el) return;

    const action = el.dataset.action;
    if(!action || !action.startsWith('playlists-')) return;

    switch(action){
      case 'playlists-create':
        return Signage.createPlaylistPrompt();

      case 'playlists-open': {
        const id = Number(el.dataset.playlistId);
        const name = el.dataset.playlistName || '';
        return Signage.openPlaylist(id, name);
      }

      case 'playlists-delete': {
        e.stopPropagation();
        const id = Number(el.dataset.playlistId);
        const name = el.dataset.playlistName || '';
        return Signage.deletePlaylist(id, name);
      }

      case 'playlists-back':
        return Signage.nav('playlists');

      case 'playlists-add-item': {
        const id = Number(el.dataset.playlistId);
        return Signage.addToPlaylist(id);
      }

      case 'playlists-remove-item': {
        const playlistId = Number(el.dataset.playlistId);
        const itemId = Number(el.dataset.itemId);
        return Signage.removeFromPlaylist(playlistId, itemId);
      }

      case 'playlists-save-fill': {
        const id = Number(el.dataset.playlistId);
        return Signage.savePlaylistFill(id);
      }
    }
  });

  // Переключатель «фиксированная длительность» — прячем/показываем поле минут
  document.addEventListener('change', e => {
    if(e.target && e.target.id === 'pl-fill-on'){
      const wrap = document.getElementById('pl-fill-minutes-wrap');
      if(wrap) wrap.style.display = e.target.checked ? '' : 'none';
    }
    if(e.target && e.target.id === 'pl-add-folder'){
      plFillMediaSelect();
    }
  });
}

// Ролики выбранной папки → во второй селект
let PL_PICK_MEDIA = [];
function plFillMediaSelect(){
  const folderSel = document.getElementById('pl-add-folder');
  const mediaSel = document.getElementById('pl-add-media');
  if(!folderSel || !mediaSel) return;
  const key = folderSel.value;
  let list = PL_PICK_MEDIA;
  if(key === 'root')             list = list.filter(m => !m.advertiser_id && !m.folder_id);
  else if(key.startsWith('f:'))  { const fid = Number(key.slice(2)); list = list.filter(m => m.folder_id === fid); }
  else if(key.startsWith('a:'))  { const aid = Number(key.slice(2)); list = list.filter(m => m.advertiser_id === aid && !m.folder_id); }
  mediaSel.innerHTML = list.length
    ? list.map(m => `<option value="${m.id}">${esc(m.title || m.filename)}</option>`).join('')
    : '<option value="">— в этой папке нет роликов —</option>';
}

async function viewPlaylists(){
  CUR_PL = null;

  document.getElementById('topright').innerHTML = canWrite()
    ? `<button class="btn primary" data-action="playlists-create">+ Плейлист</button>`
    : '';

  const view = document.getElementById('view');

  try{
    const pls = await api('/playlists');

    if(!pls.length){
      view.innerHTML = '<div class="empty">Плейлистов пока нет. Нажмите «+ Плейлист» чтобы создать.</div>';
      return;
    }

    let h = '<div class="muted" style="font-size:12px;margin-bottom:11px;">Плейлист — набор роликов, воспроизводимых по порядку. Назначается экрану через Расписание. «№» — порядковый номер (тот же номер используется при выборе плейлиста в Расписании).</div>';

    pls.forEach((p, i) => {
      const fillOn = p.fill_to_hour !== false;
      const modeBadge = fillOn
        ? `<span class="muted" style="font-size:11px;background:var(--bg2);border-radius:5px;padding:2px 8px;">блок ${Math.round((p.target_seconds||3600)/60)} мин</span>`
        : `<span class="muted" style="font-size:11px;background:var(--bg2);border-radius:5px;padding:2px 8px;">произвольный</span>`;
      h += `<div class="cell" style="display:flex;align-items:center;gap:10px;cursor:pointer;margin-bottom:7px;"
        data-action="playlists-open"
        data-playlist-id="${p.id}"
        data-playlist-name="${esc(p.name)}">
        <span style="min-width:30px;text-align:center;font-size:12px;font-weight:600;color:var(--muted);">№${i + 1}</span>
        <span style="font-size:18px;color:var(--accent);">▶</span>
        <span style="font-weight:500;flex:1;">${esc(p.name)}</span>
        ${modeBadge}
        ${canWrite() ? `<button class="btn danger" style="padding:4px 10px;font-size:12px;"
          data-action="playlists-delete"
          data-playlist-id="${p.id}"
          data-playlist-name="${esc(p.name)}">🗑</button>` : ''}
        <span class="dim">›</span>
      </div>`;
    });

    view.innerHTML = h;
  }catch(e){
    view.innerHTML = '<div class="empty">Ошибка: ' + esc(e.message) + '</div>';
  }
}

async function createPlaylistPrompt(){
  const name = prompt('Название плейлиста:');
  if(!name || !name.trim()) return;

  try{
    await api('/playlists?name=' + encodeURIComponent(name.trim()), { method:'POST' });
    toast('Плейлист создан');
    viewPlaylists();
  }catch(e){
    toast('Ошибка: ' + e.message);
  }
}

async function deletePlaylist(id, name){
  if(!confirm('Удалить плейлист «' + name + '»?')) return;

  try{
    await api('/playlists/' + id, { method:'DELETE' });
    toast('Плейлист удалён');
    viewPlaylists();
  }catch(e){
    toast('Ошибка: ' + e.message);
  }
}

async function openPlaylist(id, name){
  CUR_PL = { id, name };
  document.getElementById('topright').innerHTML = '';

  const view = document.getElementById('view');

  try{
    const [items, media, folders] = await Promise.all([
      api('/playlists/' + id + '/items'),
      api('/media'),
      api('/media/folders-all').catch(() => [])
    ]);

    let h = `<div style="display:flex;align-items:center;gap:9px;margin-bottom:13px;">
      <button class="btn" data-action="playlists-back">← Назад</button>
      <span style="font-weight:500;flex:1;">▶ ${esc(name)}</span>
    </div>
    <div id="pl-fill-panel" style="margin-bottom:13px;"></div>`;

    if(canWrite()){
      // Заглушки в выпадашке не показываем: они добавляются в блок автоматически
      PL_PICK_MEDIA = media.filter(m => !m.is_filler);
      // Папки для первого селекта: «Все», «Общая медиатека» (без рекламодателя
      // и папки), общие папки, рекламодатель без папки, папки рекламодателей
      // Показываем только папки, где есть подходящие ролики (заглушки не в счёт)
      const hasRoot = PL_PICK_MEDIA.some(m => !m.advertiser_id && !m.folder_id);
      const folderIds = new Set(PL_PICK_MEDIA.filter(m => m.folder_id).map(m => m.folder_id));
      const groups = [{key:'all', label:'Все ролики'}];
      if(hasRoot) groups.push({key:'root', label:'🗂 Общая медиатека (без папки)'});
      folders.filter(f => !f.advertiser_id && folderIds.has(f.id)).forEach(f =>
        groups.push({key:'f:'+f.id, label:'🗂 Общая / ' + f.name}));
      const advSeen = new Set();
      PL_PICK_MEDIA.forEach(m => {
        if(m.advertiser_id && !m.folder_id && !advSeen.has(m.advertiser_id)){
          advSeen.add(m.advertiser_id);
          groups.push({key:'a:'+m.advertiser_id, label:(m.advertiser||('Рекламодатель '+m.advertiser_id)) + ' (без папки)'});
        }
      });
      folders.filter(f => f.advertiser_id && folderIds.has(f.id)).forEach(f =>
        groups.push({key:'f:'+f.id, label:(f.advertiser||'') + ' / ' + f.name}));

      h += `<div style="display:flex;gap:8px;margin-bottom:14px;align-items:center;flex-wrap:wrap;">
        <select class="inp" id="pl-add-folder" style="min-width:220px;" title="Из какой папки выбирать ролик">
          ${groups.map(g => `<option value="${g.key}">${esc(g.label)}</option>`).join('')}</select>
        <select class="inp" id="pl-add-media" style="flex:1;min-width:220px;"></select>
        <input class="inp" id="pl-add-repeat" type="number" min="1" max="99" value="1" style="width:60px;" title="Повторов">
        <button class="btn primary" data-action="playlists-add-item" data-playlist-id="${id}">+ Добавить</button>
      </div>`;
    }

    if(!items.length){
      h += '<div class="empty">Плейлист пуст — добавьте ролики выше.</div>';
      view.innerHTML = h;
      plFillMediaSelect();
      loadFillPanel(id);
      return;
    }

    h += '<table style="width:100%;border-collapse:collapse;">';
    h += '<tr style="font-size:11px;color:var(--muted);"><th style="text-align:left;padding:4px 8px;">#</th><th style="text-align:left;padding:4px 8px;">Ролик</th><th style="padding:4px 8px;">Повторов</th><th style="padding:4px 8px;">Длит.</th><th></th></tr>';

    items.forEach(it => {
      h += `<tr style="border-top:0.5px solid var(--border2);">
        <td style="padding:8px;color:var(--muted);font-size:12px;">${it.position}</td>
        <td style="padding:8px;font-size:13px;">${esc(it.title || it.filename)}</td>
        <td style="padding:8px;text-align:center;font-size:12px;">${it.repeat_count}×</td>
        <td style="padding:8px;text-align:center;font-size:12px;color:var(--muted);">${it.duration_seconds ? Math.round(it.duration_seconds) + 'с' : '—'}</td>
        <td style="padding:8px;text-align:right;">
          ${canWrite() ? `<button class="iconbtn del"
            data-action="playlists-remove-item"
            data-playlist-id="${id}"
            data-item-id="${it.id}">🗑</button>` : ''}
        </td>
      </tr>`;
    });

    h += '</table>';
    view.innerHTML = h;
    plFillMediaSelect();
    loadFillPanel(id);
  }catch(e){
    view.innerHTML = '<div class="empty">Ошибка: ' + esc(e.message) + '</div>';
  }
}

// Панель «Длительность эфирного блока»: режим, факт/цель, предупреждения
async function loadFillPanel(pid){
  const box = document.getElementById('pl-fill-panel');
  if(!box) return;
  try{
    const d = await api('/playlists/' + pid + '/fill_info');
    const mins = s => (s/60).toFixed(1).replace(/\.0$/,'');
    let status;
    if(!d.fill_to_hour){
      status = `<span class="muted" style="font-size:12px;">Произвольный плейлист: ролики играют по кругу как есть — ${mins(d.main_seconds)} мин, заглушки не добавляются.</span>`;
    } else {
      const okFill = !d.warnings.length;
      status = `<span style="font-size:12px;color:${okFill?'var(--accent)':'var(--c-nike)'};">
        Основные ролики: ${d.main_plays} · ${mins(d.main_seconds)} мин.
        Заглушками добавится: ${d.filler_plays} показ(ов) · ${mins(d.filler_seconds)} мин.
        Итого блок: <b>${mins(d.total_seconds)} из ${mins(d.target_seconds)} мин</b>.
        В медиатеке заглушек: ${d.fillers_available} шт · ${mins(d.fillers_available_seconds)} мин.</span>`;
    }
    const warn = d.warnings.map(w =>
      `<div style="font-size:12px;color:var(--danger);margin-top:6px;">⚠ ${esc(w)}</div>`).join('');
    const ctl = canWrite() ? `
      <div style="display:flex;align-items:center;gap:10px;flex-wrap:wrap;margin-top:8px;">
        <label style="font-size:12px;display:inline-flex;align-items:center;gap:6px;cursor:pointer;">
          <input type="checkbox" id="pl-fill-on" ${d.fill_to_hour?'checked':''} style="width:15px;height:15px;cursor:pointer;">
          Эфирный блок фиксированной длины (добивать заглушками)</label>
        <span id="pl-fill-minutes-wrap" style="display:${d.fill_to_hour?'':'none'};align-items:center;gap:5px;">
          <input class="inp" id="pl-fill-minutes" type="number" min="1" max="1440" value="${Math.round(d.target_seconds/60)}" style="width:70px;padding:5px 8px;text-align:right;">
          <span class="muted" style="font-size:12px;">мин</span>
        </span>
        <button class="btn" style="padding:5px 12px;font-size:12px;" data-action="playlists-save-fill" data-playlist-id="${pid}">💾 Сохранить режим</button>
      </div>` : '';
    box.innerHTML = `<div class="cell">${status}${warn}${ctl}</div>`;
  }catch(e){
    box.innerHTML = `<div class="muted" style="font-size:12px;">Не удалось получить расчёт блока: ${esc(e.message)}</div>`;
  }
}

async function savePlaylistFill(pid){
  const on = !!document.getElementById('pl-fill-on')?.checked;
  const minutes = parseInt(document.getElementById('pl-fill-minutes')?.value, 10) || 60;
  try{
    await api('/playlists/' + pid + '/fill', {method:'PATCH',
      headers:{'Content-Type':'application/json'},
      body: JSON.stringify({fill_to_hour: on, target_minutes: minutes})});
    toast(on ? `Эфирный блок: ${minutes} мин` : 'Плейлист произвольный (без заглушек)');
    loadFillPanel(pid);
  }catch(e){ toast('Ошибка: ' + e.message); }
}

async function addToPlaylist(pid){
  const mid = document.getElementById('pl-add-media')?.value;
  const rc = parseInt(document.getElementById('pl-add-repeat')?.value) || 1;

  if(!mid){
    toast('Выберите ролик');
    return;
  }

  try{
    const res = await api(`/playlists/${pid}/items?media_id=${mid}&repeat_count=${rc}`, { method:'POST' });
    toast(res && res.warning ? ('Добавлен. ⚠ ' + res.warning) : 'Ролик добавлен');
    openPlaylist(pid, CUR_PL?.name || '');
  }catch(e){
    toast('Ошибка: ' + e.message);
  }
}

async function removeFromPlaylist(pid, itemId){
  if(!confirm('Удалить ролик из плейлиста?')) return;

  try{
    await api(`/playlists/${pid}/items/${itemId}`, { method:'DELETE' });
    toast('Удалено');
    openPlaylist(pid, CUR_PL?.name || '');
  }catch(e){
    toast('Ошибка: ' + e.message);
  }
}

window.Signage = window.Signage || {};
window.Signage.viewPlaylists = viewPlaylists;
window.Signage.createPlaylistPrompt = createPlaylistPrompt;
window.Signage.deletePlaylist = deletePlaylist;
window.Signage.openPlaylist = openPlaylist;
window.Signage.addToPlaylist = addToPlaylist;
window.Signage.removeFromPlaylist = removeFromPlaylist;
window.Signage.savePlaylistFill = savePlaylistFill;

initPlaylistsViewActions();

